addons/account/account_payment.py

```diff
@@ -0,0 +1,378 @@
+            raise ValidationError('The payment amount must always be positive.')
+        if self.invoice_id and self.amount > self.invoice_id.residual:
+            raise ValidationError('The amount cannot be greater than the residual amount of the invoice.')
+
+    @api.onchange('partner_type')
+    def _onchange_partner_type(self):
+        # Check partner type isn't contradictory with invoice type, if so remove the invoice
+        if self.invoice_id:
+            expected_partner_type = self.invoice_id.type in ('out_invoice', 'out_refund') and 'customer' or 'supplier'
+            if self.partner_type != expected_partner_type:
+                self.invoice_id = False
+        # Set partner_id domain
+        if self.partner_type:
+            return {'domain': {'partner_id': [(self.partner_type, '=', True)]}}
+
+    @api.onchange('payment_type')
+    def _onchange_payment_type(self):
+        # Check payment type isn't contradictory with invoice type, if so remove the invoice
+        if self.invoice_id:
+            expected_payment_type = self.invoice_id.type in ('out_invoice', 'in_refund') and 'inbound' or 'outbound'
+            if self.payment_type != expected_payment_type:
+                self.invoice_id = False
+        # Set default partner type for the payment type
+        if self.payment_type == 'inbound':
+            self.partner_type = 'customer'
+        elif self.payment_type == 'outbound':
+            self.partner_type = 'supplier'
+        # Set payment method domain
+        return self._onchange_journal()
+
+    @api.onchange('invoice_id')
+    def _onchange_invoice(self):
+        if self.invoice_id:
+            self.amount = self.invoice_id.residual
+            self.payment_type = self.invoice_id.type in ('out_invoice', 'in_refund') and 'inbound' or 'outbound'
+            self.partner_type = self.invoice_id.type in ('out_invoice', 'out_refund') and 'customer' or 'supplier'
+            self.partner_id = self.invoice_id.partner_id
+
+    @api.onchange('journal_id')
+    def _onchange_journal(self):
+        if self.journal_id:
+            # Set default payment method (we consider the first to be the default one)
+            payment_methods = self.payment_type == 'inbound' and self.journal_id.inbound_payment_methods or self.journal_id.outbound_payment_methods
+            if payment_methods:
+                self.payment_method = payment_methods[0]
+            # Hide payment method if only 'Manual' is available
+            self.hide_payment_method = len(payment_methods) == 1 and payment_methods[0].code == 'manual'
+            # Set payment method domain (restrict to methods enabled for the journal and to selected payment type)
+            payment_type = self.payment_type in ('outbound', 'transfer') and 'outbound' or 'inbound'
+            return {'domain': {'payment_method': [('type', '=', payment_type), ('id', 'in', payment_methods.ids)]}}
+
+    @api.model
+    def default_get(self, fields):
```

Comment:
remove default_get and put 'res.update({'invoice_id': self._context['invoice_id']})' in the invoice_id field definition